webrtc/README.md

```diff
@@ -161,6 +165,50 @@ MUST pass an empty string. When receiving an `RTCDataChannel` via
 an empty string. This allows future versions of this specification to make use
 of the `RTCDataChannel` `label` property.
 
+## Closing an `RTCDataChannel`
+
+Some WebRTC implementations do not guarantee that any queued messages will be
+sent after a datachannel is closed.  Other implementatations maintain separate
+outgoing message and transport queues, the status of which may not be visible
+to the user. Consequently we must add an additional layer of signaling to ensure
+reliable data delivery.
+
+When a node wishes to close a stream for writing, it SHOULD send a message with
+the `FIN` flag set, then proceed to wait for a `FIN_ACK` message from the remote
+node.
+
+If a `FIN` flag is received and the node has finished writing data, it SHOULD
+respond with a `FIN_ACK` immediately.
+
+If a `FIN` flag is received but the node has not finished writing data, it SHOULD
+delay sending a `FIN_ACK` response until it has finished writing data, sent a
+`FIN` flag, and finally received a `FIN_ACK` in response.
+
+When a `FIN_ACK` has been sent and received, a node may close the datachannel.
+
+The node MAY close the datachannel without receiving a `FIN_ACK`, for example in
+the case of a timeout, but there will be no guarantee that all previously sent
+messages have been received.
+
+If a node has previously sent a `STOP_SENDING` flag to the remote node, it MUST
+continue to act on any flags present in received messages in order to
+successfully process an incoming `FIN_ACK`.
+
+### Example of closing an `RTCDataChannel`
+
+NodeA closes for writing, NodeB delays allowing the channel to close until it also
+finishes writing.
+
+```mermaid
+sequenceDiagram
+    NodeA->>NodeB: DATA
+    NodeA->>NodeB: FIN
+    NodeB->>NodeA: DATA
+    NodeB->>NodeA: FIN
+    NodeA->>NodeB: FIN_ACK
+    NodeB->>NodeA: FIN_ACK
```

Comment:
How does this work when NodeA is a go-libp2p node and NodeB is a chromium node. 

A -> B: FIN
B -> A: Data
B -> A: FIN
A -> B: FIN_ACK
B -> A: FIN_ACK
Now as B is a chromium node the final FIN_ACK will be dropped since B is a chromium node and we don't know if the final message will be sent reliably. 
Should A just exit after a fixed period of time and exit if no FIN_ACK sort of like a timewait state?

Reply:
should we mention handling this using onClose event handler on the datachannel and to ensure checking bufferedAmount is 0 before closing?

Reply:
@achingbrain you can resolve this, if my understanding here is correct. 
specifically this:
>Since B closes the datachannel, A will be notified of the datachannel close with the onClose event handler.

Reply:
Yes, this is my understanding too.

```
1. A -> B: FIN         // A tells B it wants to close the datachannel
2. B -> A: Data        // B still sends data
3. B -> A: FIN         // B is done sending data, wants to close the datachannel
4. A -> B: FIN_ACK     // A agrees, yes, close the datachannel
5. B -> A: FIN_ACK     // B agrees, yes close the datachannel (or has closed it already?)
```

The side to close the channel is the first one to have received a `FIN_ACK` and to have sent one too (B after step 4. above).

The other side would only close the datachannel if it had sent the `FIN_ACK` to the remote (step 4 above) and the remote has not closed the channel after a timeout.

I guess then that the sending of the final `FIN_ACK` is redundant?  Can you think of a case where that would not be true?

Reply:
Sounds good.

Reply:
> > I guess then that the sending of the final FIN_ACK is redundant? Can you think of a case where that would not be true?
> 
> I can't. Any scheme which relies on the final FIN_ACK being correctly delivered can't work because chromium might drop that final message. The present scheme looks fine to me. I am also fine with keeping the final FIN_ACK as it is since it doesn't cost us much.

This is a [two-generals problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem), right? It doesn't matter how many ACKs you send, you can never be sure of the final one :)

Closing the channel can be arbitrarily interleaved, meaning we cannot depend on one party being the "first" one to close the channel. Both sides may perceive themselves as being the first one.

I think the best we can do is:

- Send `FIN` to start closing the write-half of the stream
- Keep reading from the stream
- At any point, if a node sent & received a `FIN`, close the underlying data channel
- If you receive a `FIN_ACK`, keep reading

As @sukunrt said above, what may happen is that a browser-node within Chromium receives a `FIN`, responds with `FIN_ACK`, writes some more data, sends `FIN` and as a result of the above logic immediately closes the data channel. This final `FIN` may get lost and never makes it to the remote party.

Until the chromium bug is fixed, I think we have to accept that some channels may be flagged as "closed prematurely" because we receive a `close` event without having received a `FIN` flag.

Reply:
>Chromium receives a FIN, responds with FIN_ACK, writes some more data, sends FIN and as a result of the above logic immediately closes the data channel. This final FIN may get lost and never makes it to the remote party.

This is not the speced behaviour. The behaviour in the specs is: 
```
If a FIN flag is received but the node has not finished writing data, it SHOULD
delay sending a FIN_ACK response until it has finished writing data, sent a
FIN flag, and finally received a FIN_ACK in response.
```
Which implies, 
1. Chromium receives a FIN
2. writes some more data
3. writes a FIN
4. writes a FIN_ACK 
5. waits for FIN_ACK from peer

The peer on receiving the FIN_ACK will write a FIN_ACK and close the channel, but this FIN_ACK may get lost. This is fine because the Chromium node in question will be informed of the datachannel closing by the onClose call back on the datachannel. Note: both sides would have sent a FIN and received a FIN before the onClose event handler fires. 

I must say this is a rather clever solution to the problem. @achingbrain 🙇 

Considering the discussion above, I think we should change the recommendation for delaying the FIN_ACK to MUST

Reply:
maybe the simpler solution works: see https://github.com/libp2p/specs/pull/582#discussion_r1343630055